### DISCUSSION TITLE
Redundant data safety check with fetching and error?

### DISCUSSION THREAD

**Discussion by Mafooch** [urql-graphql/urql#2894] [Q&A] [answered]
I get typechecking errors on the data object of my query result even when I first check to ensure that fetching is false and that and no error is present (I understand error and data can present concurrently).

Take this example:
```
const [{ fetching, data, error }] = useTypedQuery({
    query: { currentUser: { id: true, firstName: true } },
  });

  if (fetching) return <Typography variant="body1Bold">loading</Typography>;
  if (error) return <Typography variant="body1Bold">error</Typography>;
  return (
    <Typography variant="body1Bold">{data?.currentUser?.firstName}</Typography>
  );
```

If I don't the safety navigators in the final return, typescript gets angry. Shouldn't the combination of fetching being false and the absence of errors be sufficient for typescript to infer that data is present or am I missing something? Additionally, if we know that data is present, I would suspect that's enough to know that the actual result of my query `currentUser` is present as well, although, this may have more to do with my own implementation.

**Comment by JoviDeCroock** [comment, marked as answer]
urql holds on to the last data as well and supports partial data which means that you can have data and be fetching in the background at the same time. You can look at partial data for our normalized cache, change the variables at which point you will keep the previous data + fetch and you can use `cache-and-network` with a cache-hit to see a background fetch.

Which means that this is the expected behavior.

RE: `currentUser`, this means that your type in your backend is missing the `!` operator probably making it a nullable field

**Reply by Mafooch** [reply]
I understand the answer to the second question regarding `currentUser` but I'm still a little confused about the response to the first question.

It sounds like you're saying it's possible that data could be defined even though the query is still fetching. However, I'm still wondering how it could be possible that fetching be false, there be no errors, _and_  yet data remain undefined. You're saying that `fetching: false` isn't a `necessary` condition for data to be defined, but I'm wondering why the three aforementioned conditions are not `sufficient` for us to know that data _is_ defined.

**Reply by JoviDeCroock** [reply]
I mean, when a query starts out paused, it's fetching false without data and error 😅

**Reply by Mafooch** [reply]
Oh, right, I forgot about the paused option and how that makes that possible! I guess that answers that then. Even after checking for error and fetching, I'll still have to do a final check for data.

**Comment by DCH-DYM** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
